// @vtex/styleguide/lib/EXPERIMENTAL_Charts/helpers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getRangeOfZAxis = exports.getBarDefaultProps = exports.getLineDefaultProps = exports.getChartDefaultProps = undefined;

var _merge = require("lodash/merge");

var _merge2 = _interopRequireDefault(_merge);

var _commonProps = require("./commonProps");

var _constants = require("./LineChart/constants");

var _constants2 = require("./BarChart/constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var merge = function merge(defaultProps, userProps) {
  var props = defaultProps;
  userProps && Object.keys(userProps).forEach(function (key) {
    return props[key] = (0, _merge2.default)(props[key], userProps[key]);
  });
  return props;
};

var getRangeOfZAxis = function getRangeOfZAxis(key, data) {
  var values = data.map(function (item) {
    return item[key];
  });
  var min = Math.min.apply(Math, values);
  var max = Math.max.apply(Math, values);
  return [min, max];
};

var getChartDefaultProps = function getChartDefaultProps(userProps, customConfig) {
  if (customConfig === void 0) {
    customConfig = _commonProps.commonDefaultProps;
  }

  return {
    configs: merge(merge(_commonProps.commonDefaultProps, customConfig), userProps)
  };
};

var getLineDefaultProps = function getLineDefaultProps(userProps) {
  return {
    lineConfigs: _extends({}, _constants.defaultProps, {}, userProps)
  };
};

var getBarDefaultProps = function getBarDefaultProps(userProps) {
  return {
    barConfigs: _extends({}, _constants2.defaultProps, {}, userProps)
  };
};

exports.getChartDefaultProps = getChartDefaultProps;
exports.getLineDefaultProps = getLineDefaultProps;
exports.getBarDefaultProps = getBarDefaultProps;
exports.getRangeOfZAxis = getRangeOfZAxis;
