// @vtex/address-form/lib/postalCodeFrom/SelectLevel.js
'use strict';

exports.__esModule = true;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _AddressShapeWithValidation = require('../propTypes/AddressShapeWithValidation');

var _AddressShapeWithValidation2 = _interopRequireDefault(_AddressShapeWithValidation);

var _postalCode = require('../selectors/postalCode');

var _InputFieldContainer = require('../InputFieldContainer');

var _InputFieldContainer2 = _interopRequireDefault(_InputFieldContainer);

var _reactIntl = require('react-intl');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var SelectLevel = function (_Component) {
  _inherits(SelectLevel, _Component);

  function SelectLevel() {
    _classCallCheck(this, SelectLevel);

    return _possibleConstructorReturn(this, _Component.apply(this, arguments));
  }

  SelectLevel.prototype.render = function render() {
    var _props = this.props,
        level = _props.level,
        rules = _props.rules,
        address = _props.address,
        Input = _props.Input,
        intl = _props.intl,
        onChangeAddress = _props.onChangeAddress;

    var field = (0, _postalCode.getLevelField)(level, rules);

    return _react2.default.createElement(_InputFieldContainer2.default, {
      intl: intl,
      Input: Input,
      field: field,
      address: address,
      rules: rules,
      onChangeAddress: onChangeAddress
    });
  };

  return SelectLevel;
}(_react.Component);

SelectLevel.propTypes = process.env.NODE_ENV !== "production" ? {
  Input: _propTypes2.default.func.isRequired,
  intl: _reactIntl.intlShape,
  level: _propTypes2.default.oneOf([0, 1]),
  address: _AddressShapeWithValidation2.default,
  rules: _propTypes2.default.object.isRequired,
  onChangeAddress: _propTypes2.default.func.isRequired
} : {};

exports.default = (0, _reactIntl.injectIntl)(SelectLevel);
module.exports = exports['default'];
