// @vtex/address-form/lib/addressContainerContext.js
'use strict';

exports.__esModule = true;
exports.AddressContext = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.injectAddressContext = injectAddressContext;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var AddressContext = exports.AddressContext = _react2.default.createContext();

/* eslint-disable react/prop-types */
function injectAddressContext(Component) {
  return function AddressInjectedComponent(props) {
    if (props.address || props.onChangeAddress) return _react2.default.createElement(Component, props);

    return _react2.default.createElement(
      AddressContext.Consumer,
      null,
      function (ctx) {
        return _react2.default.createElement(Component, _extends({}, props, {
          address: ctx.address,
          onChangeAddress: ctx.handleAddressChange,
          Input: props.Input || ctx.Input
        }));
      }
    );
  };
}
