// @vtex/address-form/lib/components.js
'use strict';

exports.__esModule = true;

var _CountrySelector = require('./CountrySelector');

var _CountrySelector2 = _interopRequireDefault(_CountrySelector);

var _AddressForm = require('./AddressForm');

var _AddressForm2 = _interopRequireDefault(_AddressForm);

var _AddressSummary = require('./AddressSummary');

var _AddressSummary2 = _interopRequireDefault(_AddressSummary);

var _PostalCodeGetter = require('./PostalCodeGetter');

var _PostalCodeGetter2 = _interopRequireDefault(_PostalCodeGetter);

var _AddressContainer = require('./AddressContainer');

var _AddressContainer2 = _interopRequireDefault(_AddressContainer);

var _AutoCompletedFields = require('./AutoCompletedFields');

var _AutoCompletedFields2 = _interopRequireDefault(_AutoCompletedFields);

var _AddressRules = require('./AddressRules');

var _AddressRules2 = _interopRequireDefault(_AddressRules);

var _AddressSubmitter = require('./AddressSubmitter');

var _AddressSubmitter2 = _interopRequireDefault(_AddressSubmitter);

var _PostalCodeLoader = require('./postalCodeFrom/PostalCodeLoader');

var _PostalCodeLoader2 = _interopRequireDefault(_PostalCodeLoader);

var _GoogleMapsContainer = require('./geolocation/GoogleMapsContainer');

var _GoogleMapsContainer2 = _interopRequireDefault(_GoogleMapsContainer);

var _Map = require('./geolocation/Map');

var _Map2 = _interopRequireDefault(_Map);

var _StyleguideInput = require('./inputs/StyleguideInput');

var _StyleguideInput2 = _interopRequireDefault(_StyleguideInput);

var _StyleguideButton = require('./inputs/StyleguideButton');

var _StyleguideButton2 = _interopRequireDefault(_StyleguideButton);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  AddressContainer: _AddressContainer2.default,
  AddressForm: _AddressForm2.default,
  AddressRules: _AddressRules2.default,
  AddressSubmitter: _AddressSubmitter2.default,
  AddressSummary: _AddressSummary2.default,
  AutoCompletedFields: _AutoCompletedFields2.default,
  CountrySelector: _CountrySelector2.default,
  GoogleMapsContainer: _GoogleMapsContainer2.default,
  Map: _Map2.default,
  PostalCodeGetter: _PostalCodeGetter2.default,
  PostalCodeLoader: _PostalCodeLoader2.default,
  StyleguideInput: _StyleguideInput2.default,
  StyleguideButton: _StyleguideButton2.default
};
module.exports = exports['default'];
