// @vtex/address-form/lib/countries.js
'use strict';

exports.__esModule = true;

var _default = require('./country/default');

var _default2 = _interopRequireDefault(_default);

var _ARG = require('./country/ARG');

var _ARG2 = _interopRequireDefault(_ARG);

var _BOL = require('./country/BOL');

var _BOL2 = _interopRequireDefault(_BOL);

var _BRA = require('./country/BRA');

var _BRA2 = _interopRequireDefault(_BRA);

var _CAN = require('./country/CAN');

var _CAN2 = _interopRequireDefault(_CAN);

var _CHL = require('./country/CHL');

var _CHL2 = _interopRequireDefault(_CHL);

var _COL = require('./country/COL');

var _COL2 = _interopRequireDefault(_COL);

var _ECU = require('./country/ECU');

var _ECU2 = _interopRequireDefault(_ECU);

var _ESP = require('./country/ESP');

var _ESP2 = _interopRequireDefault(_ESP);

var _FRA = require('./country/FRA');

var _FRA2 = _interopRequireDefault(_FRA);

var _GBR = require('./country/GBR');

var _GBR2 = _interopRequireDefault(_GBR);

var _GTM = require('./country/GTM');

var _GTM2 = _interopRequireDefault(_GTM);

var _KOR = require('./country/KOR');

var _KOR2 = _interopRequireDefault(_KOR);

var _MEX = require('./country/MEX');

var _MEX2 = _interopRequireDefault(_MEX);

var _PER = require('./country/PER');

var _PER2 = _interopRequireDefault(_PER);

var _PRT = require('./country/PRT');

var _PRT2 = _interopRequireDefault(_PRT);

var _PRY = require('./country/PRY');

var _PRY2 = _interopRequireDefault(_PRY);

var _ROU = require('./country/ROU');

var _ROU2 = _interopRequireDefault(_ROU);

var _URY = require('./country/URY');

var _URY2 = _interopRequireDefault(_URY);

var _USA = require('./country/USA');

var _USA2 = _interopRequireDefault(_USA);

var _VEN = require('./country/VEN');

var _VEN2 = _interopRequireDefault(_VEN);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  defaultRules: _default2.default,
  ARG: _ARG2.default,
  BOL: _BOL2.default,
  BRA: _BRA2.default,
  CAN: _CAN2.default,
  CHL: _CHL2.default,
  COL: _COL2.default,
  ECU: _ECU2.default,
  ESP: _ESP2.default,
  FRA: _FRA2.default,
  GBR: _GBR2.default,
  GTM: _GTM2.default,
  KOR: _KOR2.default,
  MEX: _MEX2.default,
  PER: _PER2.default,
  PRT: _PRT2.default,
  PRY: _PRY2.default,
  ROU: _ROU2.default,
  URY: _URY2.default,
  USA: _USA2.default,
  VEN: _VEN2.default
}; // countries

module.exports = exports['default'];
