// @vtex/address-form/lib/inputs.js
'use strict';

exports.__esModule = true;

var _index = require('./inputs/DefaultInput/index');

var _index2 = _interopRequireDefault(_index);

var _InputError = require('./inputs/DefaultInput/InputError');

var _InputError2 = _interopRequireDefault(_InputError);

var _InputText = require('./inputs/DefaultInput/InputText');

var _InputText2 = _interopRequireDefault(_InputText);

var _InputSelect = require('./inputs/DefaultInput/InputSelect');

var _InputSelect2 = _interopRequireDefault(_InputSelect);

var _InputLabel = require('./inputs/DefaultInput/InputLabel');

var _InputLabel2 = _interopRequireDefault(_InputLabel);

var _StyleguideInput = require('./inputs/StyleguideInput');

var _StyleguideInput2 = _interopRequireDefault(_StyleguideInput);

var _GeolocationInput = require('./geolocation/GeolocationInput');

var _GeolocationInput2 = _interopRequireDefault(_GeolocationInput);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  DefaultInput: _index2.default,
  GeolocationInput: _GeolocationInput2.default,
  InputError: _InputError2.default,
  InputLabel: _InputLabel2.default,
  InputSelect: _InputSelect2.default,
  InputText: _InputText2.default,
  StyleguideInput: _StyleguideInput2.default
};
module.exports = exports['default'];
