// @vtex/address-form/lib/PostalCodeGetter.js
'use strict';

exports.__esModule = true;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _AddressShapeWithValidation = require('./propTypes/AddressShapeWithValidation');

var _AddressShapeWithValidation2 = _interopRequireDefault(_AddressShapeWithValidation);

var _constants = require('./constants');

var _OneLevel = require('./postalCodeFrom/OneLevel');

var _OneLevel2 = _interopRequireDefault(_OneLevel);

var _TwoLevels = require('./postalCodeFrom/TwoLevels');

var _TwoLevels2 = _interopRequireDefault(_TwoLevels);

var _ThreeLevels = require('./postalCodeFrom/ThreeLevels');

var _ThreeLevels2 = _interopRequireDefault(_ThreeLevels);

var _DefaultInput = require('./inputs/DefaultInput');

var _DefaultInput2 = _interopRequireDefault(_DefaultInput);

var _InputFieldContainer = require('./InputFieldContainer');

var _InputFieldContainer2 = _interopRequireDefault(_InputFieldContainer);

var _fields = require('./selectors/fields');

var _addressRulesContext = require('./addressRulesContext');

var _recompose = require('recompose');

var _addressContainerContext = require('./addressContainerContext');

var _reactIntl = require('react-intl');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var PostalCodeGetter = function (_Component) {
  _inherits(PostalCodeGetter, _Component);

  function PostalCodeGetter() {
    _classCallCheck(this, PostalCodeGetter);

    return _possibleConstructorReturn(this, _Component.apply(this, arguments));
  }

  PostalCodeGetter.prototype.render = function render() {
    var _props = this.props,
        address = _props.address,
        autoFocus = _props.autoFocus,
        loading = _props.loading,
        rules = _props.rules,
        onChangeAddress = _props.onChangeAddress,
        Input = _props.Input,
        Button = _props.Button,
        intl = _props.intl,
        onSubmit = _props.onSubmit,
        submitLabel = _props.submitLabel,
        shouldShowNumberKeyboard = _props.shouldShowNumberKeyboard;


    switch (rules.postalCodeFrom) {
      case _constants.THREE_LEVELS:
        return _react2.default.createElement(_ThreeLevels2.default, {
          intl: intl,
          loading: loading,
          Input: Input,
          Button: Button,
          address: address,
          rules: rules,
          onChangeAddress: onChangeAddress,
          onSubmit: onSubmit,
          submitLabel: submitLabel
        });
      case _constants.TWO_LEVELS:
        return _react2.default.createElement(_TwoLevels2.default, {
          intl: intl,
          loading: loading,
          Input: Input,
          Button: Button,
          address: address,
          rules: rules,
          onChangeAddress: onChangeAddress,
          onSubmit: onSubmit,
          submitLabel: submitLabel
        });
      case _constants.ONE_LEVEL:
        return _react2.default.createElement(_OneLevel2.default, {
          intl: intl,
          loading: loading,
          Input: Input,
          Button: Button,
          address: address,
          rules: rules,
          onChangeAddress: onChangeAddress,
          onSubmit: onSubmit,
          submitLabel: submitLabel
        });
      default:
      case _constants.POSTAL_CODE:
        {
          var field = (0, _fields.getField)('postalCode', rules);
          return _react2.default.createElement(_InputFieldContainer2.default, {
            intl: intl,
            loading: loading,
            Input: Input,
            Button: Button,
            field: field,
            address: address,
            autoFocus: autoFocus,
            rules: rules,
            onChangeAddress: onChangeAddress,
            onSubmit: onSubmit,
            submitLabel: submitLabel,
            shouldShowNumberKeyboard: shouldShowNumberKeyboard
          });
        }
    }
  };

  return PostalCodeGetter;
}(_react.Component);

PostalCodeGetter.defaultProps = {
  autoFocus: false,
  loading: false,
  Input: _DefaultInput2.default,
  shouldShowNumberKeyboard: false
};

PostalCodeGetter.propTypes = process.env.NODE_ENV !== "production" ? {
  address: _AddressShapeWithValidation2.default,
  autoFocus: _propTypes2.default.bool,
  loading: _propTypes2.default.bool,
  Input: _propTypes2.default.func,
  Button: _propTypes2.default.func,
  intl: _reactIntl.intlShape.isRequired,
  onChangeAddress: _propTypes2.default.func.isRequired,
  onSubmit: _propTypes2.default.func,
  rules: _propTypes2.default.object.isRequired,
  shouldShowNumberKeyboard: _propTypes2.default.bool,
  submitLabel: _propTypes2.default.string
} : {};

var enhance = (0, _recompose.compose)(_addressContainerContext.injectAddressContext, _addressRulesContext.injectRules, _reactIntl.injectIntl);
exports.default = enhance(PostalCodeGetter);
module.exports = exports['default'];
