// @vtex/address-form/lib/pureInputField.js
'use strict';

exports.__esModule = true;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _AddressShapeWithValidation = require('./propTypes/AddressShapeWithValidation');

var _AddressShapeWithValidation2 = _interopRequireDefault(_AddressShapeWithValidation);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var SHALLOW_PROPS = ['Input', 'autoFocus', 'loading', 'field', 'rules', 'options', 'onChangeAddress', 'shouldShowNumberKeyboard', 'intl'];

function shallowCompareProps(prevProps, thisProps) {
  return SHALLOW_PROPS.find(function (field) {
    return prevProps[field] !== thisProps[field];
  });
}

var SHALLOW_FIELDS = ['autoCompleted', 'loading', 'value', 'focus', 'reason', 'valueOptions', 'notApplicable', 'disabled'];

function shallowCompareField(prevProps, thisProps) {
  return SHALLOW_FIELDS.find(function (field) {
    return prevProps[field] !== thisProps[field];
  });
}

function pureInputField(WrappedComponent) {
  var PureInput = function (_Component) {
    _inherits(PureInput, _Component);

    function PureInput() {
      _classCallCheck(this, PureInput);

      return _possibleConstructorReturn(this, _Component.apply(this, arguments));
    }

    PureInput.prototype.shouldComponentUpdate = function shouldComponentUpdate(prevProps) {
      if (shallowCompareProps(prevProps, this.props)) {
        return true;
      }

      var basedOn = this.props.field.basedOn;
      if (basedOn && prevProps.address[basedOn].value !== this.props.address[basedOn].value) {
        return true;
      }

      if (shallowCompareField(prevProps.address[prevProps.field.name], this.props.address[this.props.field.name])) {
        return true;
      }

      return false;
    };

    PureInput.prototype.render = function render() {
      return _react2.default.createElement(WrappedComponent, this.props);
    };

    return PureInput;
  }(_react.Component);

  PureInput.propTypes = process.env.NODE_ENV !== "production" ? {
    autoFocus: false,
    loading: false,
    shouldShowNumberKeyboard: false
  } : {};

  PureInput.propTypes = process.env.NODE_ENV !== "production" ? {
    Input: _propTypes2.default.func.isRequired,
    autoFocus: _propTypes2.default.bool,
    loading: _propTypes2.default.bool,
    field: _propTypes2.default.object.isRequired,
    address: _AddressShapeWithValidation2.default,
    rules: _propTypes2.default.object.isRequired,
    options: _propTypes2.default.array,
    onChangeAddress: _propTypes2.default.func.isRequired,
    shouldShowNumberKeyboard: _propTypes2.default.bool
  } : {};

  return PureInput;
}

exports.default = pureInputField;
module.exports = exports['default'];
