// @vtex/styleguide/lib/InputButton/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Input = require("../Input");

var _Input2 = _interopRequireDefault(_Input);

var _withForwardedRef = require("../../modules/withForwardedRef");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

var InputButton =
/*#__PURE__*/
function (_Component) {
  _inheritsLoose(InputButton, _Component);

  function InputButton() {
    return _Component.apply(this, arguments) || this;
  }

  var _proto = InputButton.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        button = _this$props.button,
        isLoading = _this$props.isLoading,
        buttonProps = _this$props.buttonProps,
        props = _objectWithoutPropertiesLoose(_this$props, ["button", "isLoading", "buttonProps"]);

    return _react2.default.createElement("div", null, _react2.default.createElement(_Input2.default, _extends({}, props, {
      button: button,
      buttonProps: buttonProps,
      isLoadingButton: isLoading
    })));
  };

  return InputButton;
}(_react.Component);

InputButton.defaultProps = {
  buttonProps: {},
  isLoading: false
};
InputButton.propTypes = {
  /** (InputButton spec attribute) */
  button: _propTypes2.default.string,

  /** (InputButton spec attribute) */
  buttonProps: _propTypes2.default.object,

  /** Loading state */
  isLoading: _propTypes2.default.bool
};
exports.default = (0, _withForwardedRef.withForwardedRef)(InputButton);
