// @vtex/styleguide/lib/Progress/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _Steps = require("./Steps");

var _Line = require("./Line");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var propTypes = _extends({
  type: _propTypes2.default.oneOf(['line', 'steps'])
}, _Line.ProgressLineTypes, {}, _Steps.ProgressStepsTypes);

var Progress = function Progress(props) {
  if (props.type === 'line') {
    return _react2.default.createElement(_Line.ProgressLine, {
      percent: props.percent
    });
  }

  return _react2.default.createElement(_Steps.ProgressSteps, {
    steps: props.steps,
    danger: props.danger,
    slim: props.slim
  });
};

Progress.propTypes = propTypes;
Progress.defaultProps = {
  type: 'steps'
};
exports.default = Progress;
