// @vtex/address-form/lib/propTypes/AddressShapeWithValidation.js
'use strict';

exports.__esModule = true;
exports.AddressWithValidation = exports.ValidationType = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _constants = require('../constants');

var _reduce = require('lodash/reduce');

var _reduce2 = _interopRequireDefault(_reduce);

var _AddressShape = require('./AddressShape');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ValidationType = exports.ValidationType = {
  visited: false,
  valid: _propTypes2.default.bool,
  reason: _propTypes2.default.oneOf(_constants.ERROR_TYPES)
};

var AddressWithValidation = exports.AddressWithValidation = (0, _reduce2.default)(_AddressShape.Address, function (acc, valueType, propName) {
  acc[propName] = _propTypes2.default.shape(_extends({}, ValidationType, {
    autoCompleted: _propTypes2.default.bool,
    loading: _propTypes2.default.bool,
    disabled: _propTypes2.default.bool,
    value: valueType,
    valueOptions: _propTypes2.default.arrayOf(_propTypes2.default.string),
    notApplicable: _propTypes2.default.bool
  })).isRequired;
  return acc;
}, {});

exports.default = _propTypes2.default.shape(AddressWithValidation);
