// @vtex/address-form/lib/propTypes/AddressShape.js
'use strict';

exports.__esModule = true;
exports.Address = undefined;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _CountryType = require('./CountryType');

var _CountryType2 = _interopRequireDefault(_CountryType);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Address = exports.Address = {
  addressId: _propTypes2.default.string.isRequired,
  addressType: _propTypes2.default.oneOf(['residential', 'inStore', 'commercial', 'giftRegistry', 'pickup', 'search']).isRequired,
  city: _propTypes2.default.string,
  complement: _propTypes2.default.string,
  country: _CountryType2.default,
  geoCoordinates: _propTypes2.default.array,
  neighborhood: _propTypes2.default.string,
  number: _propTypes2.default.string,
  postalCode: _propTypes2.default.string,
  receiverName: _propTypes2.default.string,
  reference: _propTypes2.default.string,
  state: _propTypes2.default.string,
  street: _propTypes2.default.string,
  addressQuery: _propTypes2.default.string
};

exports.default = _propTypes2.default.shape(Address);
